src/FieldDescription/FieldDescriptionFactory.php

```diff
@@ -0,0 +1,132 @@
+        $this->guesser = $guesser;
+    }
+
+    public function create(string $class, string $name, array $options = []): FieldDescriptionInterface
+    {
+        if (!isset($options['route']['name'])) {
+            $options['route']['name'] = 'show';
+        }
+
+        if (!isset($options['route']['parameters'])) {
+            $options['route']['parameters'] = [];
+        }
+
+        [$metadata, $propertyName, $parentAssociationMappings] = $this->getParentMetadataForProperty($class, $name);
+
+        $fieldDescription = new FieldDescription(
+            $name,
+            $options,
+            $metadata->fieldMappings[$propertyName] ?? [],
+            $metadata->associationMappings[$propertyName] ?? [],
+            $parentAssociationMappings,
+            $propertyName
+        );
+
+        $guessType = $this->guesser->guess($fieldDescription);
+
+        $fieldDescription->setType($guessType->getType());
+        foreach ($guessType->getOptions() as $name => $value) {
+            if (\is_array($value)) {
+                $fieldDescription->setOption($name, array_merge($value, $fieldDescription->getOption($name, [])));
+            } else {
+                $fieldDescription->setOption($name, $fieldDescription->getOption($name, $value));
+            }
+        }
+
+        return $fieldDescription;
+    }
+
+    private function getParentMetadataForProperty(string $baseClass, string $propertyFullName): array
+    {
+        $nameElements = explode('.', $propertyFullName);
+        $lastPropertyName = array_pop($nameElements);
+        $class = $baseClass;
+        $parentAssociationMappings = [];
+
+        foreach ($nameElements as $nameElement) {
+            $metadata = $this->getMetadata($class);
+
+            if (isset($metadata->associationMappings[$nameElement])) {
+                $parentAssociationMappings[] = $metadata->associationMappings[$nameElement];
+                $class = $metadata->getAssociationTargetClass($nameElement);
+
+                continue;
+            }
+
+            break;
+        }
+
+        $properties = \array_slice($nameElements, \count($parentAssociationMappings));
+        $properties[] = $lastPropertyName;
+
+        return [
+            $this->getMetadata($class),
+            implode('.', $properties),
+            $parentAssociationMappings,
+        ];
+    }
+
+    /**
+     * @param class-string $class
+     */
+    private function getMetadata(string $class): ClassMetadata
+    {
+        return $this->getEntityManager($class)->getMetadataFactory()->getMetadataFor($class);
+    }
+
+    /**
+     * @param class-string $class
+     *
+     * @throw \RuntimeException
+     */
+    private function getEntityManager(string $class): EntityManager
+    {
+        $em = $this->registry->getManagerForClass($class);
+
+        if (!$em instanceof EntityManager) {
+            throw new \RuntimeException(sprintf('No entity manager defined for class %s', $class));
```

Comment:
IMO, if we must throw an exception based on an unmet condition from `instanceof`, the used type should be `\TypeError`.

Reply:
It's more a `if (null === $em)`check, but the instanceof is for phpstan.  `getManagerForClass` is flagged as returning a `ObjectManager|null` but this can't happen when the registry is a `ManagerRegistry`.

Reply:
I understand. Is there a criteria we can use around these checks? I mean, I saw some cases where we are using `assert()` for scenarios like this.

Reply:
Since it's possible to implement the interface with a method which return an ObjectManager, I think it's safer to keep the instanceof check.

Currently the message is more a `Entity manager not found` message.
If we throw a typeError, the message should be different.

Reply:
What I suggest is `assert($em instanceof EntityManager)`. IMO, the important thing here is to be consistent with the way we handle these cases.

Reply:
We often use exception.

Imo the rule should be
- If it's possible -> exception
- If it's not -> assert

Reply:
There are cases where there is no possible to use exceptions?
Note that I'm referring to cases where these constructions are required to make the CS tools happy, rather than the users.

Reply:
> There are cases where there is no possible to use exceptions?

I mean, there are cases like here where `!$em instanceof EntityManager` can be true, so it's better to use an exception than an assert.
And there are case like https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L114 where the assert is always true. But phpstan doesn't know this. Using an exception would be pretty useless.

Reply:
IMHO `UnexpectedValueException` is made for this.